### PR TITLE
Performance Improvements to FileTraceWriter

### LINKED ISSUE #2635: Thread exhaustion due to FileTraceWriter flush timer
One of the function app hit the thread limit issue (>512 threads).   Looking at the process dump, hundred of threads were stuck in below stack initiated by the timer.  FileTraceWriter is the same instance for almost all stacks (meaning one timer piling these threads).  The root cause is likely very slow azure file.

We suggested user to work around by setting `fileLoggingMode` to `never`.   Discussing with @fabiocav, choices of defense-in-depths could be.

- Ensure one timer work fired one after another - not blindly periodically.
- If blindly fired periodically, ensure only one is doing work.   May want to add ETW if a timer fired while work is pending to detect certain works taking too much time.

```csharp
OS Thread Id: 0x191c (334)
        Child SP               IP Call Site
0000001b8f97dd18 00007ffeaee38714 [InlinedCallFrame: 0000001b8f97dd18] Microsoft.Win32.Win32Native.GetFileAttributesEx(System.String, Int32, WIN32_FILE_ATTRIBUTE_DATA ByRef)
0000001b8f97dd18 00007ffe9ea49e1a [InlinedCallFrame: 0000001b8f97dd18] Microsoft.Win32.Win32Native.GetFileAttributesEx(System.String, Int32, WIN32_FILE_ATTRIBUTE_DATA ByRef)
0000001b8f97dcf0 00007ffe9ea49e1a DomainNeutralILStubClass.IL_STUB_PInvoke(System.String, Int32, WIN32_FILE_ATTRIBUTE_DATA ByRef)
0000001b8f97ddb0 00007ffe9e96bfb9 System.IO.File.FillAttributeInfo(System.String, WIN32_FILE_ATTRIBUTE_DATA ByRef, Boolean, Boolean)
0000001b8f97de20 00007ffe9f2940bd System.IO.FileSystemInfo.Refresh()
0000001b8f97de50 00007ffe49284cc6 Microsoft.Azure.WebJobs.Script.FileTraceWriter.SetLogFile()
0000001b8f97ded0 00007ffe4926971d Microsoft.Azure.WebJobs.Script.FileTraceWriter.Flush()
0000001b8f97df50 00007ffe9d3cb4b0 System.Timers.Timer.MyTimerCallback(System.Object)
0000001b8f97dfd0 00007ffe9e9b4750 System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
0000001b8f97e0a0 00007ffe9e9b45e4 System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
0000001b8f97e0d0 00007ffe9ea2389a System.Threading.TimerQueueTimer.CallCallback()
0000001b8f97e130 00007ffe9ea23688 System.Threading.TimerQueueTimer.Fire()
0000001b8f97e1a0 00007ffe9e9bf0fd System.Threading.QueueUserWorkItemCallback.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()
0000001b8f97e1e0 00007ffe9e9bf74e System.Threading.ThreadPoolWorkQueue.Dispatch()
0000001b8f97e678 00007ffea0b75a03 [DebuggerU2MCatchHandlerFrame: 0000001b8f97e678] 
0000001b8f97e808 00007ffea0b75a03 [ContextTransitionFrame: 0000001b8f97e808] 
0000001b8f97ea38 00007ffea0b75a03 [DebuggerU2MCatchHandlerFrame: 0000001b8f97ea38] 
OS Thread Id: 0x1d4c (335)
        Child SP               IP Call Site
0000001b8f9fdad0 00007ffeaee36704 [InlinedCallFrame: 0000001b8f9fdad0] Microsoft.Win32.Win32Native.FindFirstFile(System.String, WIN32_FIND_DATA)
0000001b8f9fdad0 00007ffe9ea51b69 [InlinedCallFrame: 0000001b8f9fdad0] Microsoft.Win32.Win32Native.FindFirstFile(System.String, WIN32_FIND_DATA)
0000001b8f9fdaa0 00007ffe9ea51b69 DomainNeutralILStubClass.IL_STUB_PInvoke(System.String, WIN32_FIND_DATA)
0000001b8f9fdbd0 00007ffe9ea3ac90 System.IO.FileSystemEnumerableIterator`1[[System.__Canon, mscorlib]].CommonInit()
0000001b8f9fdc20 00007ffe9e96873a System.IO.DirectoryInfo.InternalGetFiles(System.String, System.IO.SearchOption)
0000001b8f9fdc90 00007ffe492889cd Microsoft.Azure.WebJobs.Script.FileTraceWriter.GetLogFiles(System.IO.DirectoryInfo)
0000001b8f9fdd00 00007ffe4928889f Microsoft.Azure.WebJobs.Script.FileTraceWriter.GetCurrentLogFile()
0000001b8f9fdd40 00007ffe49284cf6 Microsoft.Azure.WebJobs.Script.FileTraceWriter.SetLogFile()
0000001b8f9fddc0 00007ffe4926971d Microsoft.Azure.WebJobs.Script.FileTraceWriter.Flush()
0000001b8f9fde40 00007ffe9d3cb4b0 System.Timers.Timer.MyTimerCallback(System.Object)
0000001b8f9fdec0 00007ffe9e9b4750 System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
0000001b8f9fdf90 00007ffe9e9b45e4 System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
0000001b8f9fdfc0 00007ffe9ea2389a System.Threading.TimerQueueTimer.CallCallback()
0000001b8f9fe020 00007ffe9ea23688 System.Threading.TimerQueueTimer.Fire()
0000001b8f9fe090 00007ffe9e9bf0fd System.Threading.QueueUserWorkItemCallback.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()
0000001b8f9fe0d0 00007ffe9e9bf74e System.Threading.ThreadPoolWorkQueue.Dispatch()
0000001b8f9fe568 00007ffea0b75a03 [DebuggerU2MCatchHandlerFrame: 0000001b8f9fe568] 
0000001b8f9fe6f8 00007ffea0b75a03 [ContextTransitionFrame: 0000001b8f9fe6f8] 
0000001b8f9fe928 00007ffea0b75a03 [DebuggerU2MCatchHandlerFrame: 0000001b8f9fe928] 
```

### PR DESCRIPTION
**Pull request by fabiocav**
Resolves #2635

### CODE REVIEW

File: src/WebJobs.Script/Diagnostics/FileTraceWriter.cs [hunk 1]
```diff
@@ -39,90 +40,141 @@ public FileTraceWriter(string logFilePath, TraceLevel level, LogType logType) :
...
         }
 
-        public override void Flush()
+        /// <summary>
+        /// Maintains a collection of objects used for locking, scoped to the file path.
+        /// The objects are not currently removed, but the list is pretty static and shouldn't be a concern.
+        /// These cannot be removed on disposal. If cleanup functionality is ever required, we can address this with a scheduled cleanup or something along those lines.
+        /// </summary>
+        /// <returns>A lock object scoped to the current <see cref="FileTraceWriter"/> log file path.</returns>
+        private object GetFlushLock() => _flushLocks.GetOrAdd(_logFilePath, new object());
```
Comment (fabiocav) [inline src/WebJobs.Script/Diagnostics/FileTraceWriter.cs]: Locks are now scoped to a given path. This change means that function and host trace writers are no longer attempting to acquire the same lock.

File: src/WebJobs.Script/Diagnostics/FileTraceWriter.cs [hunk 2]
```diff
@@ -39,90 +40,141 @@ public FileTraceWriter(string logFilePath, TraceLevel level, LogType logType) :
...
+        {
+            if (_logBuffer.Count == 0)
             {
-                // need to handle cases where log files or directories might
-                // have been deleted from underneath us
-                SetLogFile();
-                AppendToFile(_currentLogFileInfo, content);
+                if (fromTimer)
+                {
+                    StopFlushTimer();
```
Comment (fabiocav) [inline src/WebJobs.Script/Diagnostics/FileTraceWriter.cs]: As an optimization, we'll stop the timer if there is no work to do when a timer is fired. This is done here, rather than at the end, to avoid the expense of re-creating the timer more frequently.

File: src/WebJobs.Script/Diagnostics/FileTraceWriter.cs [hunk 3]
```diff
@@ -39,90 +40,141 @@ public FileTraceWriter(string logFilePath, TraceLevel level, LogType logType) :
...
             }
 
-            // check to see if we need to roll to a new log file
-            _currentLogFileInfo.Refresh();
-            if (LogFileIsOverSize(_currentLogFileInfo))
+            object flushLock = GetFlushLock();
+            if (!Monitor.TryEnter(flushLock))
             {
-                SetLogFile();
+                return false;
```
Comment (fabiocav) [inline src/WebJobs.Script/Diagnostics/FileTraceWriter.cs]: If we're unable to acquire the lock, we return immediately, without waiting and holding on to the thread.

File: src/WebJobs.Script/Diagnostics/FileTraceWriter.cs [hunk 4]
```diff
@@ -39,90 +40,141 @@ public FileTraceWriter(string logFilePath, TraceLevel level, LogType logType) :
...
-
-            // start a timer to flush accumulated logs in batches
-            _flushTimer = new Timer
-            {
-                AutoReset = true,
-                Interval = LogFlushIntervalMs
-            };
-            _flushTimer.Elapsed += OnFlushLogs;
-            _flushTimer.Start();
         }
```
Comment (fabiocav) [inline src/WebJobs.Script/Diagnostics/FileTraceWriter.cs]: The timer is no longer started at construction. We delay the timer initialization until there's something to log. If no logging is ever done, the timer will never be setup.

**Comment by fabiocav** [comment]
@suwatch FYI

File: src/WebJobs.Script/Diagnostics/FileTraceWriter.cs [hunk 5]
```diff
@@ -39,90 +40,141 @@ public FileTraceWriter(string logFilePath, TraceLevel level, LogType logType) :
...
-                AppendToFile(_currentLogFileInfo, content);
+                if (fromTimer)
+                {
+                    StopFlushTimer();
+
+                    // We just stopped our timer above; as a sanity check,
+                    // we ensure no traces were written to the buffer in the meantime.
+                    // If new traces showed up, we'll keep going and process them; otherwise,
+                    // we'll just return from here as new traces will re-enable our timer.
+                    if (_logBuffer.Count != 0)
```
Comment (mathewc) [inline src/WebJobs.Script/Diagnostics/FileTraceWriter.cs]: It isn't clear to me why you need this additional check and Start call. If the log buffer is not empty, it means Trace was called, and Trace starts the timer already, right?

File: src/WebJobs.Script/Diagnostics/FileTraceWriter.cs (hunk 3 above)
Comment (mathewc) [reply src/WebJobs.Script/Diagnostics/FileTraceWriter.cs]: Now that you've scoped the lock to a single file, if you can't enter the monitor here, it means another thread is currently flushing to the file. However isn't returning here without waiting for the lock and writing the logs potentially lossy? E.g. say an "orphaned" host that is spinning down is logging while the new host that is spinning up is also logging. We want all the logs.

File: test/WebJobs.Script.Tests/Diagnostics/FileTraceWriterTests.cs [hunk 6]
```diff
@@ -255,7 +278,7 @@ public void GetTracePrefix_ReturnsExpectedValue()
         [Fact]
         public async Task Trace_LogFileDeleted_CreatesNewFile()
         {
-            var traceWriter = new FileTraceWriter(_logFilePath, TraceLevel.Info, LogType.Host);
+            var traceWriter = new TestFileTraceWriter(_logFilePath, TraceLevel.Info, LogType.Host, true);
```
Comment (mathewc) [inline test/WebJobs.Script.Tests/Diagnostics/FileTraceWriterTests.cs]: Are we still testing the actual product now? It seems like you're overriding behavior now so tests behave differently than product?

File: test/WebJobs.Script.Tests/Diagnostics/FileTraceWriterTests.cs [hunk 7]
```diff
@@ -31,17 +31,40 @@ public FileTraceWriterTests()
...
+            {
+                if (flushed && state is TaskCompletionSource<object> tcs)
+                {
+                    tcs.SetResult(null);
+                }
+            }
+
+            int numLines = 100;
+
+            async Task WriteLogsWithState()
```
Comment (mathewc) [inline test/WebJobs.Script.Tests/Diagnostics/FileTraceWriterTests.cs]: Why these changes? Does this test still pass without these modifications - hard to understand what you're changing here.

Comment (fabiocav) [reply test/WebJobs.Script.Tests/Diagnostics/FileTraceWriterTests.cs]: These were required as the behavior the test was relying on has changed. The test expected the flush to happen even if concurrent calls were being made. That is no longer the case and those calls will return immediately when the lock isn't acquired, so this additional logic here is needed.

File: test/WebJobs.Script.Tests/Diagnostics/FileTraceWriterTests.cs (hunk 6 above)
Comment (fabiocav) [reply test/WebJobs.Script.Tests/Diagnostics/FileTraceWriterTests.cs]: I debated modifying those tests, but this is essentially replicating the behavior we'll have in production with the scheduled flushes. The problem is that these tests expect the logs to be flushed immediately after that call, but they may be delayed until the timer fires.

File: src/WebJobs.Script/Diagnostics/FileTraceWriter.cs (hunk 5 above)
Comment (fabiocav) [reply src/WebJobs.Script/Diagnostics/FileTraceWriter.cs]: This is for the small window where a race would be possible:

Thread A | Thread B
--|--
Check buffer = 0 | Tracing, hasn't written to buffer yet
***** | Acquires timer lock and starts timer
Acquires timer lock and stops timer | -

It's unlikely this would ever happen. One thing we can do is flush from when the timer is stopped.

File: src/WebJobs.Script/Diagnostics/FileTraceWriter.cs (hunk 3 above)
Comment (fabiocav) [reply src/WebJobs.Script/Diagnostics/FileTraceWriter.cs]: Yeah, that's possible. One approach would be to make explicit `Flush` calls block and wait here. That would address the issue and still avoid contention in the timer callback scenario, which was where we were seeing issues.

Comment (fabiocav) [reply src/WebJobs.Script/Diagnostics/FileTraceWriter.cs]: Making the change described above would also help remove some of the test specific code I had to add here.
